test/test_libs.py

```diff
@@ -449,6 +458,325 @@ def test_jumanji_consistency(self, envname, batch_size):
+            1,
+            transformed_out=False,
+            N=3,
+        )
+        env.reset()
+        assert not env.is_closed
+        env.rand_step()
+        assert not env.is_closed
+        env.close()
+        assert env.is_closed
+        env.reset()
+        assert not env.is_closed
+        env.close()
+
+    @pytest.mark.skipif(not torch.cuda.device_count(), reason="no cuda to test on")
+    @pytest.mark.skipif(not _has_gym, reason="no gym")
+    @pytest.mark.parametrize("frame_skip", [4])
+    @pytest.mark.parametrize("device", [0])
+    @pytest.mark.parametrize("env_name", ENVPOOL_ALL_ENVS)
+    @pytest.mark.parametrize("transformed_out", [False, True])
+    @pytest.mark.parametrize("open_before", [False, True])
+    def test_multithreaded_env_cast(
+        self,
+        env_name,
+        frame_skip,
+        transformed_out,
+        device,
+        open_before,
+        T=10,
+        N=3,
+    ):
+        # tests casting to device
+        env_multithread = _make_multithreaded_env(
+            env_name,
+            frame_skip,
+            transformed_out=transformed_out,
+            N=N,
+        )
+        if open_before:
+            td_cpu = env_multithread.rollout(max_steps=10)
+            assert td_cpu.device == torch.device("cpu")
+        env_multithread = env_multithread.to(device)
+        assert env_multithread.observation_spec.device == torch.device(device)
+        assert env_multithread.action_spec.device == torch.device(device)
+        assert env_multithread.reward_spec.device == torch.device(device)
+        assert env_multithread.device == torch.device(device)
+        td_device = env_multithread.reset()
+        assert td_device.device == torch.device(device), env_multithread
+        td_device = env_multithread.rand_step()
+        assert td_device.device == torch.device(device), env_multithread
+        td_device = env_multithread.rollout(max_steps=10)
+        assert td_device.device == torch.device(device), env_multithread
+        env_multithread.close()
+
+    @pytest.mark.skipif(not _has_gym, reason="no gym")
+    @pytest.mark.skipif(not torch.cuda.device_count(), reason="no cuda device detected")
+    @pytest.mark.parametrize("frame_skip", [4])
+    @pytest.mark.parametrize("device", [0])
+    @pytest.mark.parametrize("env_name", ENVPOOL_ALL_ENVS)
+    @pytest.mark.parametrize("transformed_out", [True, False])
+    def test_env_device(self, env_name, frame_skip, transformed_out, device):
+        # tests creation on device
+        torch.manual_seed(0)
+        N = 3
+
+        env_multithreaded = _make_multithreaded_env(
+            env_name,
+            frame_skip,
+            transformed_out=transformed_out,
+            device=device,
+            N=N,
+        )
+
+        assert env_multithreaded.device == torch.device(device)
+        out = env_multithreaded.rollout(max_steps=20)
+        assert out.device == torch.device(device)
+
+        env_multithreaded.close()
```

Comment:
We should add a simple `check_env_specs(env)` in a separate test. This will assess if the specs are ok.

Reply:
Added a separate test `test_specs`. It works for all envs except for `CheetahRun-v1`. For that env, it seems that 
the observations returned by envpool are inconsistent with the bounding box it returns in spec: the bounding box values is strictly positive, but observations can be both positive and negative. I think I'll submit this to issues in the envpool repo.

```
>>> import envpool
>>> env = envpool.make("CheetahRun-v1", env_type="gym")
>>> env.spec.observation_space
Dict(position:Box([2.22507386e-308 2.22507386e-308 2.22507386e-308 2.22507386e-308
 2.22507386e-308 2.22507386e-308 2.22507386e-308 2.22507386e-308], [1.79769313e+308 1.79769313e+308 1.79769313e+308 1.79769313e+308
 1.79769313e+308 1.79769313e+308 1.79769313e+308 1.79769313e+308], (8,), float64), velocity:Box([2.22507386e-308 2.22507386e-308 2.22507386e-308 2.22507386e-308
 2.22507386e-308 2.22507386e-308 2.22507386e-308 2.22507386e-308
 2.22507386e-308], [1.79769313e+308 1.79769313e+308 1.79769313e+308 1.79769313e+308
 1.79769313e+308 1.79769313e+308 1.79769313e+308 1.79769313e+308
 1.79769313e+308], (9,), float64))
>>> env.reset()
<TreeValue 0x7fa1cf2be8e0>
├── 'position' --> array([[-0.09139574,  0.02779732, -0.06476798, -0.01931019, -0.09387728,
│                           0.00776585, -0.00777442,  0.00666921]])
└── 'velocity' --> array([[ 0.00531753, -0.00294308,  0.00210252,  0.0133823 ,  0.01259834,
                            0.01001254, -0.00010664, -0.01476834, -0.01361733]])
```